### PR TITLE
[AMDGPU] Handle non-register operands for S_SUB/ADD_U64_PSEUDO

### PR DESCRIPTION
**Pull request by Pierre-vh**
This pseudo uses SSrc_b64 so it allows both an immediate or a register, but the lowering crashed on immediate operands.

### CODE REVIEW

**Comment by arsenm** [comment]
Missing test

**Comment by Pierre-vh** [comment]
> Missing test

How can I test ISel pseudo expansion? I didn't find the tests for it

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -4857,9 +4857,7 @@ MachineBasicBlock *SITargetLowering::EmitInstrWithCustomInserter(
     bool IsAdd = (MI.getOpcode() == AMDGPU::S_ADD_U64_PSEUDO);
     if (Subtarget->hasScalarAddSub64()) {
       unsigned Opc = IsAdd ? AMDGPU::S_ADD_U64 : AMDGPU::S_SUB_U64;
-      BuildMI(*BB, MI, DL, TII->get(Opc), Dest.getReg())
-          .add(Src0)
-          .add(Src1);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp]: Ignore clang-format, this is worse

**Comment by Pierre-vh** [comment]
> > Missing test
> 
> How can I test ISel pseudo expansion? I didn't find the tests for it

This somewhat depends on https://github.com/llvm/llvm-project/pull/79553, we currently don't emit those pseudos so it can't be tested. Either that patch needs to add a test for it, or this whole patch should be merged with it

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -4857,9 +4857,11 @@ MachineBasicBlock *SITargetLowering::EmitInstrWithCustomInserter(
     bool IsAdd = (MI.getOpcode() == AMDGPU::S_ADD_U64_PSEUDO);
     if (Subtarget->hasScalarAddSub64()) {
       unsigned Opc = IsAdd ? AMDGPU::S_ADD_U64 : AMDGPU::S_SUB_U64;
+      // clang-format off
       BuildMI(*BB, MI, DL, TII->get(Opc), Dest.getReg())
-          .addReg(Src0.getReg())
-          .addReg(Src1.getReg());
+        .add(Src0)
+        .add(Src1);
+      // clang-format on
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:4864]: Adding these comments is worse than putting up with clang-format's preferred indentation.

**Comment by jayfoad** [comment]
> but the lowering crashed on register operands

Crashed on **non**-register operands?

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:4864]: Drop the comments and ignore the error. I really wish clang-format would just ignore anything involving BuildMI

**Comment by arsenm** [comment]
> > Missing test
> 
> How can I test ISel pseudo expansion? I didn't find the tests for it

Ideally there would be a test that generated the failing pseudo in the first place, but you can also just run finalize-isel on MIR

**Comment by Pierre-vh** [comment]
> > but the lowering crashed on register operands
> 
> Crashed on **non**-register operands?

Oops, I meant immediate

File: llvm/test/CodeGen/AMDGPU/add_sub_u64_pseudos.mir [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1100 -run-pass=finalize-isel -verify-machineinstrs -o - %s | FileCheck -check-prefix=GFX11 %s
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1200 -run-pass=finalize-isel -verify-machineinstrs -o - %s | FileCheck -check-prefix=GFX12 %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/add_sub_u64_pseudos.mir:3]: the verify is redundant?

**Review by arsenm — APPROVED** [review]
A test that really selected this way would be better

**Pull request merged**